File: src/gdt.c
#include "gdt.h"

sd_t               g_sdes[16];
cpu_state_t       *g_cpu_state; // In the future should be an array of CPU
extern void        stack_top(void);

static inline void iomap_set(uint16_t port)
{
    uint32_t index = IOMAP_INDEX(port);
    uint8_t  bit = IOMAP_BIT(port);
    g_cpu_state->io_bitmap[index] |= (1 << bit);
}

static inline void iomap_clear(uint16_t port)
{
    uint32_t index = IOMAP_INDEX(port);
    uint8_t  bit = IOMAP_BIT(port);
    g_cpu_state->io_bitmap[index] &= ~(1 << bit);
}

static inline uint8_t iomap_test(uint16_t port)
{
    uint32_t index = IOMAP_INDEX(port);
    uint8_t  bit = IOMAP_BIT(port);
    return (g_cpu_state->io_bitmap[index] & (1 << bit)) ? 1 : 0;
}

static __always_inline void write_tss_entry(void)
{
    // save stack segment selector;
    g_cpu_state->tss.ss0 = 0x24;

    // call of stack_top to save kernel stack address
    g_cpu_state->tss.esp0 = (uint32_t)stack_top;

    // Authorize 0x3d4 and 0x3d5 port
    iomap_clear(0x3d4);
    iomap_clear(0x3d5);
}

static __always_inline void load_tss(void)
{
    asm volatile(".intel_syntax noprefix\n\t"
                 "mov ax, 0x40 \n\t"
                 "ltr ax\n\t"
                 ".att_syntax prefix\n\t"
                 :
                 :
                 : "ax", "memory");
}

static inline uint8_t
create_type(uint8_t exec, uint8_t dc, uint8_t rw, uint8_t access)
{
    return access | rw << 1 | dc << 2 | exec << 3;
}

static inline uint8_t create_flags(uint8_t gran, uint8_t mode, uint8_t avl)
{
    return avl | 0 << 1 | mode << 2 | gran << 3;
}

static inline uint8_t create_access(uint8_t type, uint8_t system, uint8_t dpl)
{
    return type | system << 4 | dpl << 5 | 1 << 7;
}

static inline void create_descriptor(uint8_t  flags,
                                     uint8_t  access,
                                     uint32_t addr,
                                     uint32_t limit,
                                     sd_ptr   segment_des)
{
    segment_des->start = (limit & 0xffff) | (addr & 0xffff) << 16;
    limit >>= 16;
    addr >>= 16;

    uint8_t low_address = (addr & 0xff);
    addr >>= 8;

    uint8_t high_address = (addr & 0xff);
    addr >>= 8;

    segment_des->end = low_address | (access & 0xff) << 8 | (limit & 0xf) << 16
                     | (flags & 0xf) << 20 | high_address << 24;
}

// reload only kernel descriptor other descriptor are loaded in gdtr but not
// used.
static __always_inline void reload_segments(void)
{
    __asm__ volatile(".intel_syntax noprefix\n\t"
                     "push 0x08\n\t"
                     "push offset 1f\n\t"
                     "lret\n"
                     "1:\n\t"
                     "mov ax, 0x10\n\t"
                     "mov ds, ax\n\t"
                     "mov es, ax\n\t"
                     "mov fs, ax\n\t"
                     "mov ax, 0x18\n\t"
                     "mov ss, ax\n\t"
                     "mov ax, 0x20\n\t"
                     "mov gs, ax\n\t"
                     ".att_syntax prefix\n\t"
                     :
                     :
                     : "ax", "memory");
}

static __always_inline void load_gdt(void)
{
    struct gdt gdt;
    gdt.size = sizeof(g_sdes) - 1;
    gdt.address = (uint32_t)&g_sdes;

    asm volatile("cld\n\t"
                 "mov %0, %%esi\n\t"
                 "mov %1, %%edi\n\t"
                 "mov %2, %%ecx\n\t"
                 "rep movsb\n\t"
                 :
                 : "r"(&g_sdes), "r"(0x800), "r"(sizeof(g_sdes))
                 : "esi", "edi", "ecx", "memory");
    asm volatile("lgdt %0\n\t" ::"m"(gdt));
}

static inline void test_user_mode_function(void)
{
    printk("Test in usermode\n");
    printk("CPL: %d\n", get_current_level_privilege());
    while (1) {
    }
}

__attribute__((naked, noreturn)) void jump_usermode(void)
{
    asm volatile(".intel_syntax noprefix\n\t"
                 "mov ax, 0x30 | 3\n\t"
                 "mov ds, ax\n\t"
                 "mov es, ax\n\t"
                 "mov fs, ax\n\t"
                 "push 0x30 | 3\n\t"
                 "push esp\n\t"
                 "pushf\n\t"
                 "push 0x28 | 3\n\t"
                 "push %0\n\t"
                 "iret\n\t"
                 ".att_syntax prefix\n\t"
                 :
                 : "r"(test_user_mode_function)
                 : "ax", "memory");
}

// TODO: Need to handle SMP.
void init_cpu_state(void)
{
    // zero tss and disable all io ports.
    memset(&g_cpu_state->tss, 0, sizeof(tss_segment_t));
    memset(g_cpu_state->io_bitmap, 0xFF, sizeof(g_cpu_state->io_bitmap));
    g_cpu_state->tss.iomap_base = (uint16_t)offsetof(cpu_state_t, io_bitmap);
    g_cpu_state->end_marker = 0xFF;
    write_tss_entry();
}

void init_gdt(void)
{
    // NULL DESCRIPTOR
    create_descriptor(0, 0, 0, 0, &g_sdes[0]);

    // KERNEL CODE DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_ON, TYPE_DC_OFF, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_0),
        0x00000000,
        0x000FFFFF,
        &g_sdes[1]);

    // KERNEL DATA DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_OFF, TYPE_DC_OFF, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_0),
        0x00000000,
        0x000FFFFF,
        &g_sdes[2]);

    // KERNEL STACK DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_OFF, TYPE_DC_ON, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_0),
        0x00000000,
        0x000FFFFF,
        &g_sdes[3]);

    // create_descriptor(0, 0, 0, 0, &g_sdes[4]);
    // VGA DESCRIPTOR (this is only present so you can write string)
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_OFF, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_OFF, TYPE_DC_OFF, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_3),
        0x000B8000,
        0x00000FFF,
        &g_sdes[4]);

    // USER CODE DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_ON, TYPE_DC_OFF, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_3),
        0x00000000,
        0x000FFFFF,
        &g_sdes[5]);

    // USER DATA DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_OFF, TYPE_DC_OFF, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_3),
        0x00000000,
        0x000FFFFF,
        &g_sdes[6]);

    // USER STACK DESCRIPTOR
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_ON, FLAGS_MODE_ON, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_OFF, TYPE_DC_ON, TYPE_RW_ON, TYPE_A_OFF),
            ACCESS_DESCRIPTOR_TYPE_ON,
            ACCESS_DPL_RING_3),
        0x00000000,
        0x000FFFFF,
        &g_sdes[7]);

    // TASK STATE DESCRIPTOR
    uint32_t base = (uint32_t)&g_cpu_state->tss;
    uint32_t limit = (uint32_t)&g_cpu_state->end_marker - base + 1;
    create_descriptor(
        create_flags(FLAGS_GRANULARITY_OFF, FLAGS_MODE_OFF, FLAGS_AVL_64_OFF),
        create_access(
            create_type(TYPE_EXEC_ON, TYPE_DC_OFF, TYPE_RW_OFF, TYPE_A_ON),
            ACCESS_DESCRIPTOR_TYPE_OFF,
            ACCESS_DPL_RING_0),
        base,
        limit,
        &g_sdes[8]);

    load_gdt();
    load_tss();
    reload_segments();
}

File: src/gdt/gdt.zig
const std = @import("std");
const console = @import("../vga/vga.zig");
const mem = @import("../lib/mem.zig");

comptime {
    _ = mem;
}

/// GdtFlags contains all the flags that are not Types or DPL for segment descriptors.
const GdtFlags = enum(u32) {
    System = 1 << 12, // 12 bit system descriptor
    Present = 1 << 15, // 15 bit present descriptor (if set in memory)
    Accessed = 1 << 20, // 20 bit available for use by system
    DefaultBig = 1 << 22, // 22 bit default operation size (if set 32-bit segment)
    Granularity = 1 << 23, // 23 bit granularity flag
};

/// Decsriptores Types for the 8-11 bit inside segment descriptor.
const DescriptorTypes = enum(u32) {
    Accessed = 1 << 8, // 8 bit access bit for segment
    Readable = 1 << 9, // 9 bit write bit for segment
    Expend = 1 << 10, // 10 bit expend for data segment (expand down grow stack down)
    Data = 1 << 11, // 11 bit data segment (if set to 0 data) else code
};

/// For Segment Descriptor privilege level needed to be used.
const DescriptorPrivilegeLevel = enum(u3) {
    Ring_0 = 0,
    Ring_1,
    Ring_2,
    Ring_3,
};

const SegmentDescriptor = packed struct {
    start: u32,
    end: u32,
};

export var g_sdes: [16]SegmentDescriptor  = [_]SegmentDescriptor{.{ .start = 0, .end = 0 }} ** 16;

inline fn createDescriptorPrivilege(lvl: DescriptorPrivilegeLevel) u32 {
    const lvl_int: u32 = @intFromEnum(lvl);
    return lvl_int << 13;
}

inline fn createSegmentDescriptor(base: u32, limit: u32, flags: u32, types: u32, lvl: DescriptorPrivilegeLevel, segs: *SegmentDescriptor) void {
    //const state = struct {
    //    var idx: usize = 0;
    //};
    segs.start = (limit & 0xffff) | (base & 0xffff) << 16;
    const new_limit = limit >> 16;
    var new_base = base >> 16;

    const low_address: u32 = (new_base & 0xff);
    new_base >>= 8;

    const high_address: u32 = (new_base & 0xff) << 24;
    new_base >>= 8;

    const limit_value: u32 = (new_limit & 0xf) << 16;

    segs.end = low_address | high_address | limit_value | types | flags | createDescriptorPrivilege(lvl);
    console.print("test\n", .{});
    //state.idx += 1;
}


inline fn reloadSegments() void {
    asm volatile(
        \\  ljmp $0x08, $1f 
        \\1:
        \\  mov $0x10, %%ax
        \\  mov %%ax, %%ds
        \\  mov %%ax, %%es
        \\  mov %%ax, %%fs
        \\  mov %%ax, %%gs
        \\  mov $0x20, %%ax 
        \\  mov %%ax, %%ss
        ::: .{ .ax = true, .memory = true });
}

const Gdt = packed struct {
    size: u16,
    addrs: u32,
};

inline fn loadGdt() void {
    var gdt: Gdt = .{ .size = 0, .addrs = 0 };
    gdt.size = (g_sdes.len * @sizeOf(SegmentDescriptor)) - 1;
    gdt.addrs = @intFromPtr(&g_sdes);
    //console.print("gdt addrs: {*} gdt.addrs: 0x{x}\n", .{&gdt, gdt.addrs});
    asm volatile (
        \\ cli
        \\ lgdt %[addr]
        :
        : [addr] "m" (gdt),
        : .{ .memory = true }
    );
    // reloadSegments();
}


pub fn init() void {
    // NULL DECSRIPTOR
    for (0..16) |i| {
        console.print("[{d}]: g_sdes.start=0x{x} g_sdes.end=0x{x}\n", .{i, g_sdes[i].start, g_sdes[i].end});
    }

    var flags: u32 = @intFromEnum(GdtFlags.Granularity) | @intFromEnum(GdtFlags.Present) | @intFromEnum(GdtFlags.System) | @intFromEnum(GdtFlags.DefaultBig);
    var types: u32 = @intFromEnum(DescriptorTypes.Data) | @intFromEnum(DescriptorTypes.Readable);
    // KERNEL CODE DESCRIPTOR
    createSegmentDescriptor(0x00000000, 0xFFFFFFFF, flags, types, .Ring_0, &g_sdes[1]);

    flags = @intFromEnum(GdtFlags.Granularity) | @intFromEnum(GdtFlags.Present) | @intFromEnum(GdtFlags.System) | @intFromEnum(GdtFlags.DefaultBig);
    types = @intFromEnum(DescriptorTypes.Readable);
    // KERNEL DATA DESCRIPTOR
    createSegmentDescriptor(0x00000000, 0xFFFFFFFF, flags, types, .Ring_0, &g_sdes[2]);

    flags = @intFromEnum(GdtFlags.Granularity) | @intFromEnum(GdtFlags.System) | @intFromEnum(GdtFlags.Present) | @intFromEnum(GdtFlags.DefaultBig);
    types = @intFromEnum(DescriptorTypes.Readable) | @intFromEnum(DescriptorTypes.Expend);
    // KERNEL STACK DESCRIPTOR
    createSegmentDescriptor(0x00000000, 0xFFFFFFFF, flags, types, .Ring_0, &g_sdes[3]);
    for (0..16) |i| {
        console.print("[{d}]: g_sdes.start=0x{x} g_sdes.end=0x{x}\n", .{i, g_sdes[i].start, g_sdes[i].end});
    }
    loadGdt();
}
